### PR TITLE
Https Redirect Middleware

### PR DESCRIPTION
**Pull request by mikaelm12**
Adding a convenience `RedirectToHttps` middleware.
For issue: https://github.com/aspnet/BasicMiddleware/issues/240

@javiercn @danroth27 @DamianEdwards @Tratcher

### CODE REVIEW

File: BasicMiddleware.sln [hunk 1]
```diff
@@ -1,10 +1,13 @@
 Microsoft Visual Studio Solution File, Format Version 12.00
 # Visual Studio 15
-VisualStudioVersion = 15.0.26127.3
+VisualStudioVersion = 15.0.26510.0
 MinimumVisualStudioVersion = 10.0.40219.1
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.AspNetCore.HttpOverrides", "src\Microsoft.AspNetCore.HttpOverrides\Microsoft.AspNetCore.HttpOverrides.csproj", "{517308C3-B477-4B01-B461-CAB9C10B6928}"
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "src", "src", "{A5076D28-FA7E-4606-9410-FEDD0D603527}"
+	ProjectSection(SolutionItems) = preProject
+		src\Microsoft.AspNetCore.Rewrite\RewriteMiddleware.cs = src\Microsoft.AspNetCore.Rewrite\RewriteMiddleware.cs
```
Comment (Tratcher) [inline BasicMiddleware.sln]: ???

File: samples/RewriteSample/Startup.cs [hunk 2]
```diff
@@ -33,10 +39,22 @@ public static void Main(string[] args)
...
-                        // Configure SSL
+                        // Configure HTTPS
                         listenOptions.UseHttps("testCert.pfx", "testPassword");
                     });
                 })
+                .ConfigureAppConfiguration((builderCtx, configurationBuilder) =>
+                {
+                    configurationBuilder.AddInMemoryCollection(new Dictionary<string, string>
+                    {
+                        ["Microsoft:AspNetCore:RedirectToHttps:Port"] = "5001"
```
Comment (Tratcher) [inline samples/RewriteSample/Startup.cs]: Why not just add a real config file to the sample? This kind of hack only makes sense in a test.

File: samples/RewriteSample/Startup.cs [hunk 3]
```diff
@@ -33,10 +39,22 @@ public static void Main(string[] args)
...
+                    });
+                })
+                .ConfigureServices(services =>
+                {
+                    services.AddTransient<IConfigureOptions<RedirectToHttpsOptions>, ConfigureDefaults<RedirectToHttpsOptions>>();
```
Comment (Tratcher) [inline samples/RewriteSample/Startup.cs]: Comment that this is normally handled by Microsoft.AspNetCore.dll

File: src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToHttps/RedirectToHttpsMarkerService.cs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Rewrite
+{
+    class RedirectToHttpsMarkerService
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToHttps/RedirectToHttpsMarkerService.cs]: Needs a comment explaining its usage.

File: src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToHttps/RedirectToHttpsServiceCollectionExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+using Microsoft.Extensions.DependencyInjection.Extensions;
+using Microsoft.Extensions.Options.Infrastructure;
+
+namespace Microsoft.AspNetCore.Rewrite
+{
+    public static class RedirectToHttpsServiceCollectionExtensions
+    {
+        public static IServiceCollection AddRedirectToHttps(this IServiceCollection services)
+        {
+            services.AddOptions();
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToHttps/RedirectToHttpsServiceCollectionExtensions.cs]: hosting adds options, you shouldn't need to.

File: src/Microsoft.AspNetCore.Rewrite/RedirectToHttpsExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        {
+            if (app == null)
+            {
+                throw new ArgumentNullException(nameof(app));
+            }
+
+            var marker = app.ApplicationServices.GetService<RedirectToHttpsMarkerService>();
+            if (marker == null)
+            {
+                throw new InvalidOperationException("The RedirectToHttps service needs to be registered");
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/RedirectToHttpsExtensions.cs]: or options need to be specified

File: src/Microsoft.AspNetCore.Rewrite/RedirectToHttpsExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            var marker = app.ApplicationServices.GetService<RedirectToHttpsMarkerService>();
+            if (marker == null)
+            {
+                throw new InvalidOperationException("The RedirectToHttps service needs to be registered");
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/RedirectToHttpsExtensions.cs]: Why do you need this if options have been specified?

File: src/Microsoft.AspNetCore.Rewrite/RedirectToHttpsMiddleware.cs [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+using Microsoft.Extensions.FileProviders;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Rewrite
+{
+    public class RedirectToHttpsMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly IFileProvider _fileProvider;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/RedirectToHttpsMiddleware.cs]: unused

File: src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToHttps/RedirectToHttpsOptionsSetup.cs [hunk 9]
```diff
@@ -0,0 +1,16 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Options.Infrastructure;
+
+namespace Microsoft.AspNetCore.Rewrite
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToHttps/RedirectToHttpsOptionsSetup.cs]: .Internal? @HaoK

File: src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToHttps/RedirectToHttpsServiceCollectionExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+            services.TryAddSingleton<RedirectToHttpsMarkerService>();
+            services.TryAddEnumerable(ServiceDescriptor.Singleton<ConfigureDefaultOptions<RedirectToHttpsOptions>,RedirectToHttpsOptionsSetup>());
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToHttps/RedirectToHttpsServiceCollectionExtensions.cs]: https://github.com/aspnet/KestrelHttpServer/commit/c6e228d1768c96f8a416f1e4886af3ef18eb1b17#diff-db54d93f809f6d8bd20dd571e30b1262R35

File: src/Microsoft.AspNetCore.Rewrite/RedirectToHttpsMiddleware.cs [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.FileProviders;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Rewrite
+{
+    public class RedirectToHttpsMiddleware
+    {
+        private readonly RequestDelegate _next;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/RedirectToHttpsMiddleware.cs]: unused

File: test/Microsoft.AspNetCore.Rewrite.Tests/RedirectToHttpsMiddlewareTests.cs [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+            .ConfigureServices(services =>
+            {
+                services.AddTransient<IConfigureOptions<RedirectToHttpsOptions>, ConfigureDefaults<RedirectToHttpsOptions>>();
+                services.AddRedirectToHttps();
+            });
+            var server = new TestServer(builder);
+
+            var response = await server.CreateClient().GetAsync(new Uri("http://example.com:1234"));
+
+            Assert.Equal("https://example.com:443/", response.Headers.Location.OriginalString);
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.Rewrite.Tests/RedirectToHttpsMiddlewareTests.cs]: The redirect should not be adding the port if it's the default 443 for https.

File: src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToHttps/RedirectToHttpsOptionsSetup.cs (hunk 9 above)
Comment (HaoK) [inline src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToHttps/RedirectToHttpsOptionsSetup.cs]: Yep, put this in Microsoft.AspNetCore.Rewrite.Internal

File: src/Microsoft.AspNetCore.Rewrite/RedirectToHttpsExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+            return app.UseMiddleware<RedirectToHttpsMiddleware>();
+        }
+
+        /// <summary>
+        /// Ensures the scheme is Https
+        /// </summary>
+        /// <param name="app">The <see cref="IApplicationBuilder"/></param>
+        /// <param name="options">Options for the redirect to Https.</param>
+        /// <returns></returns>
+        public static IApplicationBuilder UseRedirectToHttps(this IApplicationBuilder app, RedirectToHttpsOptions options)
```
Comment (HaoK) [inline src/Microsoft.AspNetCore.Rewrite/RedirectToHttpsExtensions.cs]: UseHttpsRedirect/HttpsRedirectOptions?  Sounds a little bit cleaner to me...

File: src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToHttps/RedirectToHttpsServiceCollectionExtensions.cs [hunk 14]
```diff
@@ -0,0 +1,20 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+using Microsoft.Extensions.Options.Infrastructure;
+
+namespace Microsoft.AspNetCore.Rewrite
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/RedirectToHttps/RedirectToHttpsServiceCollectionExtensions.cs]: Microsoft.Extensions.DependencyInjection

**Pull request closed without merging**